### ISSUE TITLE
Semantics of IOperation.IsInvalid

### ISSUE THREAD

**Issue by mavasani** [dotnet/roslyn#18089]
Currently, this flag seems to report true if any of the descendant bound nodes within Bound node for the given IOperation has any error. For example, consider the below code, where the block and switch statement are completely valid, but there is an invalid case label.

```
using System;

class Program
{
    static void Main(string[] args)
    {
        var x = new Program();
        switch (x.ToString())
        {
            case 1:
                break;
        }
    }
}
```

Now look at the IOperation tree for the method body:
```
IOperation tree for "{
        var x = new Program();
        switch (x.ToString())
        {
            case 1:
                break;
        }
    }"

IBlockStatement (2 statements, 1 locals) (OperationKind.BlockStatement, IsInvalid)
  Local_1: Program x
  IVariableDeclarationStatement (1 variables) (OperationKind.VariableDeclarationStatement)
    IVariableDeclaration: Program x (OperationKind.VariableDeclaration)
      Initializer: IObjectCreationExpression (Constructor: Program..ctor()) (OperationKind.ObjectCreationExpression, Type: Program)
  ISwitchStatement (1 cases) (OperationKind.SwitchStatement, IsInvalid)
    Switch expression: IInvocationExpression (virtual System.String System.Object.ToString()) (OperationKind.InvocationExpression, Type: System.String)
        Instance Receiver: ILocalReferenceExpression: x (OperationKind.LocalReferenceExpression, Type: Program)
    ISwitchCase (1 case clauses, 1 statements) (OperationKind.SwitchCase, IsInvalid)
      Case clauses: ISingleValueCaseClause (Equality operator kind: BinaryOperationKind.StringEquals) (CaseKind.SingleValue) (OperationKind.SingleValueCaseClause, IsInvalid)
          IConversionExpression (ConversionKind.Invalid, Implicit) (OperationKind.ConversionExpression, Type: System.String, IsInvalid)
            ILiteralExpression (Text: 1) (OperationKind.LiteralExpression, Type: System.Int32, Constant: 1)
      Body: IBranchStatement (BranchKind.Break) (OperationKind.BranchStatement)
```

The invalid case label seems to cause containing ISwitchStatement and IBlockStatement to have `IsInvalid = true`, which seems counter-intuitive as the switch and block statements are not semantically erroneous, but they contain a descendant node with semantic errors. I would have expected the property to be either called `HasErrors` OR ensure the `IsInvalid` property is true only if the current IOperation is invalid, not one of the descendants.

**Comment by jinujoseph** [comment]
**Design Team Decision** 

we should remove the invalid API. Diagnostics for the ioperation can be derived from the semantic model .if we see a pattern of usage then we can add helper method for it .

**Comment by bkoelman** [comment]
Using test-driven development for [my `IOperation`-based analyzers](https://github.com/bkoelman/CSharpGuidelinesAnalyzer), life was great. Only when I started dogfooding, many `NullReferenceException`s started to surface due to broken code during typing in the IDE. To overcome this, I [added a check](https://github.com/bkoelman/CSharpGuidelinesAnalyzer/blob/b5bffab636fb3f741158af3df44b79e68772a285/src/CSharpGuidelinesAnalyzer/CSharpGuidelinesAnalyzer/Extensions/AnalysisContextExtensions.cs#L50) for `.IsInvalid` at the start of all analyzers.

Hope this helps in finding a pattern of usage. I'd like my analyzers to remain silent on (any) invalid code.

**Comment by heejaechang** [comment]
@bkoelman every Ioperation has syntax or you can use semantic model (http://source.roslyn.io/#Microsoft.CodeAnalysis/Compilation/SemanticModel.cs,379) to see whether you are dealing with invalid code.

we could provide an extension method for people but you can easily write one yourself based on what you believe by "invalid"

**Comment by bkoelman** [comment]
An extension method would be fine I guess. Could you provide an example for the descendant invalid case label at the top of this issue? I wonder how to get that from syntax or the semantic model.

**Comment by heejaechang** [comment]
invalid == ioperation.syntax.GetDiagnostics().Any() || semanticMode.GetDiagnostics(ioperation.syntax.Span()).Any()

something like this.

**Comment by bkoelman** [comment]
Thanks. I'm now using the following:

```csharp
public static bool IsInvalid(this IOperation operation, Compilation compilation)
{
    SemanticModel model = compilation.GetSemanticModel(operation.Syntax.SyntaxTree);
    return operation.IsInvalid(model);
}

public static bool IsInvalid(this IOperation operation, SemanticModel model)
{
    bool hasSyntaxErrors = operation.Syntax.GetDiagnostics().Any(IsError);
    bool hasSemanticErrors = model.GetDiagnostics(operation.Syntax.Span).Any(IsError);

    return hasSyntaxErrors || hasSemanticErrors;
}

private static bool IsError(Diagnostic diagnostic)
{
    return diagnostic.Severity == DiagnosticSeverity.Error;
}
```

Which has similar performance in my tests, compared to the `IOperation.IsInvalid` property. Compilation and model are probably cached.

**Comment by heejaechang** [comment]
it is nit. but since checking semantic error is more expensive then semantic, I would bail out if syntax error exist before checking semantic error.

if (hasSyntax...) return true;
bool hasSemanticErrors = ...
if (hasSemantic ..) return true;

...

also, we are working on to make sure all IOperation to have Syntax, but right now, there can be some bug where Syntax == null, so to make it safe, I would do this as well

if (operation.Syntax == null) return true;

thank you

**Comment by bkoelman** [comment]
Thanks, changed to the following:

```csharp
public static bool IsInvalid(this IOperation operation, Compilation compilation)
{
    if (operation.Syntax != null)
    {
        bool hasSyntaxErrors = operation.Syntax.GetDiagnostics().Any(IsError);
        if (!hasSyntaxErrors)
        {
            SemanticModel model = compilation.GetSemanticModel(operation.Syntax.SyntaxTree);

            bool hasSemanticErrors = model.GetDiagnostics(operation.Syntax.Span).Any(IsError);
            if (!hasSemanticErrors)
            {
                return false;
            }
        }
    }

    return true;
}

private static bool IsError(Diagnostic diagnostic)
{
    return diagnostic.Severity == DiagnosticSeverity.Error;
}
```

Are there open issues for the cases where Syntax == null? Perhaps I should add some more tests.

**Comment by jinujoseph** [comment]
fixed with https://github.com/dotnet/roslyn/pull/20436

**Comment by heejaechang** [comment]
closed in favor of https://github.com/dotnet/roslyn-analyzers/pull/1243

**Comment by bkoelman** [comment]
@heejaechang The extension method in #20436 does not contain the check on syntax first, like you suggested above. Is that intentional? 

Also I missed the `CancellationToken`, which I'll add to my version.

**Comment by heejaechang** [comment]
@bkoelman looks like SemanticModel.GetDiagnostics checks syntax one as well.

http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp/Compilation/CSharpCompilation.cs,2304
